cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/DistributedSchedulerService.java

```diff
@@ -65,11 +66,22 @@ public void onChange(ServiceDiscovered serviceDiscovered) {
           if (!Iterables.isEmpty(serviceDiscovered) && !schedulerStarted.get()) {
             LOG.info("Starting scheduler, Discovered {} dataset service(s)",
                      Iterables.size(serviceDiscovered));
-            try {
-              startScheduler();
-              schedulerStarted.set(true);
-            } catch (Throwable t) {
-              LOG.error("Exception when starting scheduler.", t);
+            int retries = 0;
+            while (true) {
+              try {
+                startScheduler();
+                schedulerStarted.set(true);
+                LOG.info("Scheduler service started successfully.");
+                break;
+              } catch (Throwable t) {
+                retries++;
+                LOG.warn("Error during retry# {} - {}", retries, t);
```

Comment:
should it be ``t.getMessage`` or something? Otherwise I'm not sure how this would work. We shouldn't log the entire stack trace for sure.